File: files/300-399/384/CH-384.py
import pandas as pd

path = "300-399/384/CH-384 Table Transformation.xlsx"
df = pd.read_excel(path, usecols="B:E", skiprows=2, nrows=8)
test = pd.read_excel(path, usecols="G:J", skiprows=2, nrows=6)

name = df.iloc[::2].reset_index(drop=True)
sale = df.iloc[1::2].reset_index(drop=True)

name["Date"] = name["Date"].ffill()

result = (
    pd.concat([
        name[["Date", "Customer"]].assign(Product=name["product 1"], Sale=sale["product 1"]),
        name[["Date", "Customer"]].assign(Product=name["product 2"], Sale=sale["product 2"]),
    ])
    .dropna()
    .sort_index(kind="stable")
    .reset_index(drop=True)
)

result["Sale"] = result["Sale"].astype(int)

test.columns = result.columns
print(result.equals(test))

File: files/300-399/384/CH-384.R
library(tidyverse)
library(readxl)

path <- "300-399/384/CH-384 Table Transformation.xlsx"
df <- read_excel(path, range = "B3:E11", col_names = TRUE)
test <- read_excel(path, range = "G3:J9", col_names = TRUE)

name <- df[seq(1, nrow(df), 2), ]
sale <- df[seq(2, nrow(df), 2), ]
name <- name |>
  mutate(Date = zoo::na.locf(Date))
result <- bind_rows(
  name |>
    transmute(
      Date,
      Customer,
      Product = `product 1`,
      Sale = sale$`product 1`
    ),
  name |>
    transmute(Date, Customer, Product = `product 2`, Sale = sale$`product 2`)
) |>
  mutate(.row = rep(seq_len(nrow(name)), 2)) |>
  drop_na() |>
  arrange(.row) |>
  select(-.row) |>
  mutate(Sale = as.integer(Sale))

colnames(test) <- colnames(result)

all.equal(result, test, check.attributes = FALSE)
# [1] TRUE
